### PR TITLE
Avoid silencing relevant follow-up errors

### PR DESCRIPTION
**Pull request by oli-obk**
r? @matthewjasper

This PR only adds new errors to tests that are already failing and fixes one ICE.

Several tests were changed to not emit new errors. I believe all of them were faulty tests, and not explicitly testing for the code that had new errors.

### CODE REVIEW

**Review by matthewjasper — COMMENTED** [review]
Nice to see someone working on this.

There are a couple of cases where I think the new additional error messages are not helpful. It would be good to either have them fixed in this PR or have a clear proposal to fix them.

File: tests/ui/rfcs/rfc-2093-infer-outlives/self-structs.rs [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 #[rustc_outlives]
 struct Foo<'a, 'b, T> { //~ ERROR rustc_outlives
+    //~^ ERROR rustc_outlives
```
[the remark this answers is not in the document]
Comment (oli-obk) [reply tests/ui/rfcs/rfc-2093-infer-outlives/self-structs.rs:5]: This is just a compiler debugging tool that I think doesn't matter if the UX is weird

**Comment by matthewjasper** [comment]
Fair enough. r=me with the comment in `rustc_hir_analysis/src/lib.rs` restored

**Comment by apiraino** [comment]
At a quick glance, it seems this is waiting for a rebase and then it's ready for merge, am I right? Switching to waiting on author. 

@rustbot author

File: compiler/rustc_hir_analysis/src/lib.rs [hunk 2]
```diff
@@ -168,12 +168,9 @@ pub fn check_crate(tcx: TyCtxt<'_>) -> Result<(), ErrorGuaranteed> {
 
     // this ensures that later parts of type checking can assume that items
     // have valid types and not error
-    // FIXME(matthewjasper) We shouldn't need to use `track_errors`.
```
[the remark this answers is not in the document]
Comment (oli-obk) [reply compiler/rustc_hir_analysis/src/lib.rs:169]: tracking this now via https://github.com/rust-lang/rust/issues/117514

**Comment by matthiaskrgr** [comment]
Hmm, looks like this causes ICEs that were previously hidden by errors. :/

**Comment by oli-obk** [comment]
imo that's a good thing. We were always playing with fire by using `track_errors`

I'll happily do more ICE squashing.

**Pull request merged**